src/main/scala/firrtl/Compiler.scala

```diff
@@ -208,47 +208,53 @@ abstract class Transform extends TransformLike[CircuitState] with DependencyAPI[
 
   private def emptySet = new scala.collection.mutable.LinkedHashSet[TransformDependency]
 
-  override def prerequisites = inputForm match {
+  override def prerequisites: Seq[Dependency] = inputForm match {
     case C => Nil
     case H => Forms.Deduped
     case M => Forms.MidForm
     case L => Forms.LowForm
     case U => Nil
   }
 
-  override def optionalPrerequisites = inputForm match {
+  override def optionalPrerequisites: Seq[Dependency] = inputForm match {
     case L => Forms.LowFormOptimized
     case _ => Seq.empty
   }
 
-  override def dependents = {
-    val lowEmitters = classOf[LowFirrtlEmitter] :: classOf[VerilogEmitter] :: classOf[MinimumVerilogEmitter] ::
-      classOf[SystemVerilogEmitter] :: Nil
+  override def dependents: Seq[Dependency] = {
+    val lowEmitters = DependencyID[LowFirrtlEmitter] :: DependencyID[VerilogEmitter] :: DependencyID[MinimumVerilogEmitter] ::
+      DependencyID[SystemVerilogEmitter] :: Nil
 
     val emitters = inputForm match {
-      case C => classOf[ChirrtlEmitter]      :: classOf[HighFirrtlEmitter]   :: classOf[MiddleFirrtlEmitter] :: lowEmitters
-      case H => classOf[HighFirrtlEmitter]   :: classOf[MiddleFirrtlEmitter] :: lowEmitters
-      case M => classOf[MiddleFirrtlEmitter] :: lowEmitters
+      case C => DependencyID[ChirrtlEmitter]      :: DependencyID[HighFirrtlEmitter]   :: DependencyID[MiddleFirrtlEmitter] :: lowEmitters
+      case H => DependencyID[HighFirrtlEmitter]   :: DependencyID[MiddleFirrtlEmitter] :: lowEmitters
+      case M => DependencyID[MiddleFirrtlEmitter] :: lowEmitters
       case L => lowEmitters
       case U => Nil
     }
 
+    val selfDep = DependencyID.fromTransform(this)
+
     inputForm match {
-      case C => (emptySet ++ Forms.VerilogOptimized                           ++ emitters - this.getClass).toSeq
-      case H => (emptySet ++ Forms.VerilogOptimized -- Forms.Deduped          ++ emitters - this.getClass).toSeq
-      case M => (emptySet ++ Forms.VerilogOptimized -- Forms.MidForm          ++ emitters - this.getClass).toSeq
-      case L => (emptySet ++ Forms.VerilogOptimized -- Forms.LowFormOptimized ++ emitters - this.getClass).toSeq
+      case C => (Forms.VerilogOptimized.toSet                           ++ emitters - selfDep).toSeq
+      case H => (Forms.VerilogOptimized.toSet -- Forms.Deduped          ++ emitters - selfDep).toSeq
+      case M => (Forms.VerilogOptimized.toSet -- Forms.MidForm          ++ emitters - selfDep).toSeq
+      case L => (Forms.VerilogOptimized.toSet -- Forms.LowFormOptimized ++ emitters - selfDep).toSeq
```

Comment:
I don't think this PR can remove the `emptySet` for dependents. This needs to stay ordered and going from `Seq => Set => Seq` won't preserve order.

(However, I learned all this from your `DiGraph` code, @albert-magyar, so I assume you had some reason for doing this?)

Reply:
Good point, I must have just missed the fact that `emptySet` made an ordered set. I'll change this back.